Feat: Improve multiline input handling in popups

Wolfsblvt opened the pull request:
Not sure this was intended in the past before or wasn't. There were some other checks on when to actually use "Send as Enter" in popups.
But when I tested working with popups, I felt it was quite unintuitive. If a popup has a single-line input as the main input, then pressing enter should send and commit the popup. Makes sense.
If the input is multiple and text easily being editable with linebreaks, it should **not** commit and send the input simply because Enter is pressed. That would feel very counter-intuitive.

Still kept a short-hand to send without clicking the button, but a more sane keybind.

### Changes

- Added Ctrl+Enter requirement for submission in multiline input popups to prevent accidental sends
- Exported PopupUtils class for external use

## Copilot Summary
This pull request improves the user experience for multiline input popups and makes a utility class available for import elsewhere. The most important changes are grouped below.

Multiline input behavior improvements:

* Updated the `Popup` class in `public/scripts/popup.js` so that pressing Enter in a multiline input does not automatically send the input unless the Ctrl key is held, making interactions more intuitive for users.

Codebase modernization:

* Changed `PopupUtils` from a regular class to an exported class in `public/scripts/popup.js`, allowing it to be imported and used in other modules.

<!-- Put X in the box below to confirm -->

## Checklist:

- [x] I have read the [Contribution guidelines](https://www.youtube.com/watch?v=e-GwojpaoQI).

Wolfsblvt committed:
feat: improve multiline input handling in popups

- Added Ctrl+Enter requirement for submission in multiline input popups to prevent accidental sends
- Exported PopupUtils class for external use

Wolfsblvt commented:
@codex review

Cohee1207 commented:
Can we gather the list of all multiline input popups in the current codebase to estimate the impact? I'm sure there is a non-zero possibility to reenact xkcd workflow with this one if we affect some common input dialog.

Wolfsblvt commented:
> Can we gather the list of all multiline input popups in the current codebase to estimate the impact? I'm sure there is a non-zero possibility to reenact xkcd workflow with this one if we affect some common input dialog.

Should be easy enough. There aren't too many.

Thanks, codex integration

1. [public/script.js#L7268-L7271](https://github.com/SillyTavern/SillyTavern/blob/staging/public/script.js#L7268-L7271) – Onboarding popup requests a username with `rows: 2`. @public/script.js#7268-7271  
2. [public/script.js#L11624-L11627](https://github.com/SillyTavern/SillyTavern/blob/staging/public/script.js#L11624-L11627) – External character import popup uses `rows: 4`. @public/script.js#11624-11627  
3. [public/scripts/scrapers.js#L173-L176](https://github.com/SillyTavern/SillyTavern/blob/staging/public/scripts/scrapers.js#L173-L176) – Web scrape URL prompt sets `rows: 4`. @public/scripts/scrapers.js#173-176  
4. [public/scripts/scrapers.js#L528-L532](https://github.com/SillyTavern/SillyTavern/blob/staging/public/scripts/scrapers.js#L528-L532) – YouTube scrape prompt sets `rows: 2`. @public/scripts/scrapers.js#528-532  
5. [public/scripts/personas.js#L421-L431](https://github.com/SillyTavern/SillyTavern/blob/staging/public/scripts/personas.js#L421-L431) – Persona description popup uses `rows: 4`. @public/scripts/personas.js#421-431  
6. [public/scripts/extensions/caption/index.js#L107-L114](https://github.com/SillyTavern/SillyTavern/blob/staging/public/scripts/extensions/caption/index.js#L107-L114) – Caption refinement popup displays `rows: 5`. @public/scripts/extensions/caption/index.js#107-114  
7. [public/scripts/extensions/caption/index.js#L306-L311](https://github.com/SillyTavern/SillyTavern/blob/staging/public/scripts/extensions/caption/index.js#L306-L311) – Custom caption prompt popup sets `rows: 2`. @public/scripts/extensions/caption/index.js#306-311  
8. [public/scripts/extensions/stable-diffusion/index.js#L732-L736](https://github.com/SillyTavern/SillyTavern/blob/staging/public/scripts/extensions/stable-diffusion/index.js#L732-L736) – Stable Diffusion prompt refinement popup uses `rows: 5`. @public/scripts/extensions/stable-diffusion/index.js#732-736  
9. [public/scripts/extensions/connection-manager/index.js#L286-L292](https://github.com/SillyTavern/SillyTavern/blob/staging/public/scripts/extensions/connection-manager/index.js#L286-L292) – New connection profile naming popup sets `rows: 2`. @public/scripts/extensions/connection-manager/index.js#286-292  
10. [public/scripts/extensions/connection-manager/index.js#L607-L611](https://github.com/SillyTavern/SillyTavern/blob/staging/public/scripts/extensions/connection-manager/index.js#L607-L611) – Profile rename popup also sets `rows: 2`. @public/scripts/extensions/connection-manager/index.js#607-611  
11. [public/scripts/slash-commands.js#L3528-L3532](https://github.com/SillyTavern/SillyTavern/blob/staging/public/scripts/slash-commands.js#L3528-L3532) – Slash command input popup defaults `rows` to 4 when unspecified. @public/scripts/slash-commands.js#3528-3532  
12. [data/default-user/extensions/Extension-WebSearch/index.js#L1517-L1529](https://github.com/SillyTavern/SillyTavern/blob/staging/data/default-user/extensions/Extension-WebSearch/index.js#L1517-L1529) – API key entry popup uses `rows: 2`. @data/default-user/extensions/Extension-WebSearch/index.js#1517-1529  
13. [public/scripts/extensions/third-party/SillyTavern-Pronouns/replacer.js#L230-L234](https://github.com/SillyTavern/SillyTavern/blob/staging/public/scripts/extensions/third-party/SillyTavern-Pronouns/replacer.js#L230-L234) – Pronoun replacer popup configures `rows: 8`. @public/scripts/extensions/third-party/SillyTavern-Pronouns/replacer.js#230-234

(Man, I love finding actually effective and totally useful use-cases of AI agents)

Wolfsblvt commented:
Looking at the list, I think the only slightly debatable ones are the URL ones. Web scrape and character import. Where multiline is a supported feature, but often enough people just want to paste one URL and confirm.
I *could* see that breaking the flow for people.

Still, I think this feature is more sensible. Otherwise, we'll just have to add another one to the dozen existing popup options...

Cohee1207 commented:
I think every place with two rows will be needed to be reduced to 1 to support confirm on Enter. Anything larger is more of a defined text area where you'd expect Enter to do a line break.

Wolfsblvt commented:
I would be okay with that. So changing the argument for all popups with `rows: 2` to `rows: 1` and keeping the implementation?

Cohee1207 commented:
Yep, except I'd maybe change this one to 4:

> [public/scripts/extensions/caption/index.js#L306-L311](https://github.com/SillyTavern/SillyTavern/blob/staging/public/scripts/extensions/caption/index.js#L306-L311) – Custom caption prompt popup sets rows: 2. @public/scripts/extensions/caption/index.js#306-311

Wolfsblvt committed:
refactor: remove redundant higher/different rows from input popups

- Removed rows: 2 from callGenericPopup calls where default behavior is sufficient
- Increased rows from 2 to 4 in caption extension for better multiline input experience

Wolfsblvt commented:
Done.

Cohee1207 reviewed (approved):
Okay, let's see how it goes. It's not like there's an awful lot of multiline inputs in popups anyway, as I see.

Wolfsblvt commented:
I think it's mostly slash command popups that'll benefit from that, that ask people to enter some kind of prompt or something via a QR button.
And extensions.
